apm-agent-plugins/apm-okhttp-plugin/src/main/java/co/elastic/apm/agent/okhttp/OkHttpClientAsyncInstrumentation.java

```diff
@@ -0,0 +1,166 @@
+        callbackWrapperCreator = HelperClassManager.ForAnyClassLoader.of(tracer,
+            OkHttpClientAsyncInstrumentation.class.getName() + "$CallbackWrapperCreator",
+            OkHttpClientAsyncInstrumentation.class.getName() + "$CallbackWrapperCreator$CallbackWrapper");
+    }
+
+    @VisibleForAdvice
+    public static class OkHttpClient3ExecuteAdvice {
+
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        private static void onBeforeEnqueue(@Advice.Origin Class<? extends Call> clazz,
+                                            @Advice.FieldValue(value = "originalRequest", typing = Assigner.Typing.DYNAMIC, readOnly = false) @Nullable Request originalRequest,
+                                            @Advice.Argument(value = 0, readOnly = false) @Nullable Callback callback,
+                                            @Advice.Local("span") Span span) {
+            if (tracer == null || tracer.getActive() == null || callbackWrapperCreator == null) {
+                return;
+            }
+
+            if (originalRequest == null || callback == null) {
+                return;
+            }
+
+            final TraceContextHolder<?> parent = tracer.getActive();
+
+            Request request = originalRequest;
+            span = HttpClientHelper.startHttpClientSpan(parent, request.method(), request.url().toString(), request.url().getHost());
+            if (span != null) {
+                final WrapperCreator<Callback> wrapperCreator = callbackWrapperCreator.getForClassLoaderOfClass(clazz);
+                if (wrapperCreator != null) {
+                    span.activate().markLifecycleManagingThreadSwitchExpected();
+                    originalRequest = originalRequest.newBuilder().addHeader(TraceContext.TRACE_PARENT_HEADER, span.getTraceContext().getOutgoingTraceParentHeader().toString()).build();
+                    callback = wrapperCreator.wrap(callback, span);
+                }
+            }
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class)
+        private static void onAfterEnqueue(@Advice.Local("span") @Nullable Span span) {
+            if (span != null) {
+                span.deactivate();
+            }
+        }
+    }
+
+    public static class CallbackWrapperCreator implements WrapperCreator<Callback> {
+
+        @Override
+        public Callback wrap(final Callback delegate, Span span) {
+            return new CallbackWrapper(span, delegate);
+        }
+
+        private static class CallbackWrapper implements Callback {
+            private final Span span;
+            private final Callback delegate;
+
+            CallbackWrapper(Span span, Callback delegate) {
+                this.span = span;
+                this.delegate = delegate;
+            }
+
+            @Override
+            public void onFailure(Request req, IOException e) {
+                try {
+                    span.captureException(e).end();
```

Comment:
Is `onResponse` called after this as well? If not- can you find the response code in `e` maybe?

Reply:
I think in those cases there is no response, for example, because of a network error. If there is an error status code, `onResponse` would be called.